### ISSUE TITLE
App not launching

### ISSUE THREAD

**Issue by jontshelton** [EmilyBjoerk/lsml#743]
**OS:** Windows 10
**LSML version:** 2.0.9
**Screen resolution:** 1920x1080
**Java version:** 8u171
**MWO Dir** E:\SteamLibrary\steamapps\common\MechWarrior Online

**Steps to reproduce issue:**
1. attempt to Open LSML

**Actual result:**
Error: This Application requires a Java Runtime Environment 1.8.171

**Expected result:**
Open LSML

**Additional steps taken**
Remove/restart/reinstall java
Update to newer java version
Open .jar manually (opens but missing button labels, see screenshot)

<?xml version="1.0" encoding="UTF-8"?>

<!DOCTYPE properties SYSTEM "http://java.sun.com/dtd/properties.dtd">

-<properties>

<comment>Written by LSML</comment>

<entry key="ui_pgiCompatibility">true</entry>

<entry key="core_lastUpdateCheck">1547351771849</entry>

<entry key="core_gameInstallDir">E:\SteamLibrary\steamapps\common\MechWarrior Online</entry>

<entry key="ui_useCompactLayout">false</entry>

<entry key="ui_useSmallMechList">true</entry>

<entry key="core_garageFile">C:\Users\Admin\Documents\mwogarage.lsxml</entry>

</properties>

![screenshot 1](https://user-images.githubusercontent.com/46641137/51082065-76dfdc80-16cd-11e9-9e22-d1e61e097e33.png)

**Comment by EmilyBjoerk** [comment]
Likely it is a problem with the exe launcher (launch4j) it seems as there have been many recent problems with it's detection of Java with the newer versions (9,10,11?). 

Launching it from the jar directly as you did should work just normally. What you're seeing is that it hasn't loaded the CSS file for the UI, which is odd as it is packed into the JAR.

Can you try uninstalling any other versions of Java that are newer than Java 8 to confirm that this is a problem with the exe launcher?

Can you also open the jar (it's just a zip file renamed to jar) and look under resources/view to see if there are any .css/.bss  files there? Should be a hand full.

**Comment by Smokey2** [comment]
The Java update killed mine too.

**Comment by EmilyBjoerk** [comment]
Huh... that's odd. There should be CSS files in there as well. I changed that for the 2.0.9 release, see: commit b4a96b9903cfeeb005cd4b4e2e51241b17c602ff which is in 2.0.9 unless I messed up the release badly. 

I did test 2.0.9 on my Win 10 machine (as I do all releases) and it worked with the latest Java 8 version at the time. So it *should* work with the stated java version (of course I will update it to work with newer ones in the next patch).

Do you have Java on your PATH variable or do you have JAVA_HOME set?

2.0.10 is currently in early access for Patreons, I think it should resolve your issue when it releases publicly in a week. If not I will put out a hotfix after it is out.

**Comment by jontshelton** [comment]
Hmm OK then if there is a new release on the way maybe it is best to just wait and see if that fixes it.

There is a Java PATH, but HOME was not set as I have been using only the JRE not the full JDK. I just installed the JDK and checked the variables, but am getting the same errors.

If there are files missing from the install, could there be a problem with the msi? Something on my end causing it to not function correctly, or a missing dependency that your test machine already has from an earlier release?

**Comment by EmilyBjoerk** [comment]
I think it is possible that launch4j looks at your PATH first if there is a Java entry there. So my thought was that this could affect a) the Java version error and b) the missing css styles due to actually running on another JRE than you think after installing the right Java version. JRE/JDK doesn't matter.

The msi installer, just copies the jar. Either your download got corrupted, which is not so likely with checksums and the lack of other errors; Or I somehow messed up making the jar which seems more likely. It's not a missing dependency everything is packed into the jar and ran from it, and you would see other errors if they were.

What I think happened is that JavaFX used to support Binary CSS (bss) and JavaFX used to be included in the JRE between versions 5-8 (not sure about the earliest version). However with the release of Java 9 the tool to create the BSS files got killed and I think with Java 10 (or 11?) the JavaFX library got extracted from the JRE to a [separate library](https://openjfx.io). As LSML used BSS to speed up loading on slow machines and was written initially targeting JavaFX 7/8 it seems as Oracle is "moving fast and breaking things", or at least my program (╯°□°）╯︵ ┻━┻. 

Leading up to 2.0.9 I tried to migrate to Java 9 (#714) but ran into issues. Tooling didn't seem ready and it was confusing how to get JavaFX working with the modularisation and there seemed to be little to no documentation on what was happening with JavaFX, or at least I didn't find it. I decided to stick to Java 8 for the time, however the tool to create the BSS files disappeared in a recent update to Java 8, prompting that commit I linked earlier. Now (ignoring the fact that I messed up the package and the css files I intended to be there, aren't) it seems from your screenshot that your JRE can't load the BSS files. When I tested the 2.0.9 release, my JRE could render it just fine from the jar so I must surmise that this BSS feature was unceremoniously killed in an update to Java 8 (or higher) recently as well (you said you installed 8 and uninstalled all other versions right?). 

It does seem like things are stabilising a bit with JavaFX, being in a separate library is ultimately a good thing as it can serve to isolate it from the whirlwind of breakage that seems to be Oracle's new development ideas. So I might look into trying the migration again for the next big release.

**Comment by jontshelton** [comment]
I only use java for this so I didn't have any other versions installed anyways. When I first installed LSML I downloaded the current java 8 release, and when that didn't work I uninstalled it and downloaded 171 off of CNET or somewhere similar. When that didn't work I tried a few other versions, as well as switching from 64bit to 32bit. Then I tried google to see if anyone else had similar issues, but the few results I got were related to a much older release. I ran through the common java troubleshooting pages, then I came here.

Incidentally, I fired up my old PC, which is also windows 10, but has an old Java 7 version of LSML on it, and it still works.

Is it possible the error tooltip is asking for the wrong java version? It doesn't seem to work with 171 or 201, but it worked when you tested it, which would imply that there is some "magic release" somewhere between the two that is new enough to support the app but old enough to not have been broken yet.

**Comment by EmilyBjoerk** [comment]
Hmm...

So I finally got around to testing on my development machine and I installed 2.0.9 and ran it (I have Win 10 Pro and Java 1.8.0_172 (not 171) and 10.0.1 installed. It detected my Java just fine, reads the BSS files and renders normally. What the deuce.

For the record my release process is to make sure all unit tests pass and make sure that the 64 bit version will install cleanly and start up without MWO installed on my dev machine that has the minimum supported Java version installed.

**Comment by EmilyBjoerk** [comment]
I don't make launch4j which builds the exe. But I find it unlikely that the tooltip is wrong. And 171 is the version I have specified here: https://github.com/EmilyBjoerk/lsml/blob/53cb5dd276c8b8f8ce94ad436afc4417aca06d43/build.gradle#L115

**Comment by EmilyBjoerk** [comment]
I get the same behaviour as you in a VM with a fresh Win 10 and 8u172 which is the same as on my dev machine.

But why does it work on my dev machine...? And more concerning, if it doesn't work on the minimum version on a fresh machine and it's been out for 5 months with over 400 downloads... Why am I only hearing of this now? And I know for certain of people using it just fine... Did 2.0.9 work previously for you?

**Comment by EmilyBjoerk** [comment]
Ok, I think I have some answers...

* Java 11 will *NOT RUN ANY CURRENT LSML VERSION AT ALL* due to JavaFX no longer being a part of Oracle JRE/JDK. And also Oracle is not releasing a JRE for Java11, instead developers are [expected to build and bundle their own minimal JREs with jlink](https://medium.com/azulsystems/using-jlink-to-build-java-runtimes-for-non-modular-applications-9568c5e70ef4) and use [OpenJFX](https://openjfx.io).
* Java 9/10 will run 2.0.9 and 2.0.10 correctly in all aspects.
* Java 9/10 are no longer available for download from Oracle. - DO'H!
  * Correction, just have to dig a little deeper: https://www.oracle.com/technetwork/java/javase/downloads/java-archive-javase10-4425482.html 
* Java 8u172 and later (and possibly all the way back to 8u77, see commit: 1cc2631b8240dc624344c234cc3717cbcc107461) will not read the BSS files in LSML 2.0.4 - 2.0.9 (see commit d8c511504bec4a8ab2f8bb7b7e840d37ff5e1a1b). 
* The exe launcher will incorrectly detect Java 8 JRE version, but correctly detect Java 9/10 versions (wtf?)
  * Both for 2.0.9 and 2.0.10.
* LSML 2.0.10 has the CSS files packaged but the exe launcher still says wrong java version but the jar runs correctly if manually launched.
* LSML 2.0.9 is missing the CSS files because I made the release before I tagged in git so the tag was younger than the release. Incorrectly showing the commit to remove bss as included when it was not.


**I will repackage LSML 2.0.10 as 2.0.11 to solve the launcher problem.** The version bump is necessary to make uninstall/upgrade of msi not go crazy. LSML will target Java 8 for the time being and migrate to Java 11 and bundled JREs eventually, but I'm loathe to have to build architecture dependent binaries as opposed to a jar that runs everywhere.

# Speculative flow of events
The exe launcher in 2.0.8 wouldn't run without Java 9 due to my botched attempted migration to Java 9. However Java 9/10 JREs weren't publicly promoted for download (See #725) but I'm assuming people knew how to google and found and downloaded Java 9/10 on their own (or just gave up on LSML, oops).

Newer Java 8 versions removed the BSS support, but because of my botched migration attempt any remaining, existing users were already on Java 9/10 and 2.0.9 worked fine for them. New users were either on Java 9/10 as well or just gave up on LSML. In September 2018 Java 10 med End of Life and  soon after users were forcibly "upgraded"(?) to a later Java 8 version (https://www.java.com/en/download/more_info10.jsp) or the update just removed Java 9/10.

As people are eventually hitting that "Java updates are available" button in the system tray LSML starts "breaking" as users are put on Java 8 again.

**Comment by jontshelton** [comment]
That sounds about right, I never installed java when I built this PC, and 9/10 was gone by the time I got around to it.

I've done tech support, and the average user doesn't even read the error message before they close it, much less do any effective troubleshooting. I'm sure most people either gave up immediately or after updating java didn't help.

Anyways, thanks for figuring things out. I was starting to think there was a problem on my end preventing java from working properly.
